Adding intial working version of checks ci server

rachmari opened the pull request:
This updates the `server.rb` file to perform authentication, webhook signature verification, and payload conversion (to JSON) in `helpers` methods called by the `before` filter. This creates a template where the entire route can be implemented in a quickstart. We explain what this code does in the updates made to the "Building your first GitHub App" quickstart before having readers modify code. 

The `advanced_server.rb` file sets up check runs and also performs a CI test using Rubocop. The messages that Rubocop reports are annotated in the Checks tab. A "Fix it" button is displayed that will automatically correct errors that Rubocop can fix and submits the changes to the PR.

DEGoodmanWilson commented on advanced_server.rb:
```diff
@@ -0,0 +1,314 @@
+      clone_repository(full_repo_name, repository, head_sha)
+
+      @report = `rubocop '#{repository}/*' --format json` # --auto-correct`
+      `rm -rf #{repository}`
```
To be fair, I didn't do any better 😹 
https://github.com/DEGoodmanWilson/excellent-ci

rachmari replied:
@DEGoodmanWilson 🙈 I didn't see that code until just now. Our ecosystem docs issue was referencing this repo: `DEGoodmanWilson/checks-api-ruby-demo`. Is looks like simplabs/excellent is much easier to use, and we could use the GitHub Git API with it to avoid command line completely. It looks like the only advantage of using to Rubcop versus excellent is that Rubocop can demonstrate the Checks API actions feature because it supports `--auto-correct`.

DEGoodmanWilson replied:
RuboCop is better-known, and I _love_ the addition of the actions feature. Let's go with what you have.

rachmari commented on advanced_server.rb:
```diff
@@ -0,0 +1,314 @@
+            end_column   = offense["location"]["last_column"]
+            message      = offense["message"]
+            annotation = {
+              "filename" => file_path,
```
👍

rachmari commented on advanced_server.rb:
```diff
@@ -0,0 +1,314 @@
+        @report = `rubocop '#{repository}/*' --format json --auto-correct`
+        pwd = Dir.getwd()
+        Dir.chdir(repository)
+        `git add .`
```
@DEGoodmanWilson that's an excellent point. I'll look into that!

rachmari commented on advanced_server.rb:
```diff
@@ -0,0 +1,314 @@
+        Dir.chdir(repository)
+        `git add .`
+        `git commit -am 'Automatically fix Octo Rubocop notices.'`
+        `git push 'https://github.com/#{full_repo_name}.git' #{head_branch}`
```
@DEGoodmanWilson I really wanted to find a linter that I could avoid command line calls with, but I didn't have any luck. 

> I feel like there is a sequence of API calls that we should be using here to ensure this actually succeeds.

Are you referring to the https://developer.github.com/v3/git/ API? Originally, I planned to use this API but didn't think it would work when running the linter on the command line.

swinton replied:
You'll need to authenticate as your installation for this to work: https://developer.github.com/apps/building-github-apps/authenticating-with-github-apps/#http-based-git-access-by-an-installation

DEGoodmanWilson replied:
The app will already have authenticated as an installation, that's no problem. The question I think @swinton is how can we `git push` code from disk using that installation key?

rachmari replied:
@swinton @DEGoodmanWilson I've already authenticated when I clone the repo using [this](https://developer.github.com/apps/building-github-apps/authenticating-with-github-apps/#http-based-git-access-by-an-installation) method. Do I still need to authenticate when I push code?

swinton replied:
I'm not :100:, I _think_ you'll need to provide the token again when you push, i.e. 

```ruby
`git push 'https://x-access-token:#{@installation_token.to_s}@github.com/#{full_repo_name}.git' #{head_branch}`
```

rachmari replied:
I don't believe the access token works on push. I reached out in slack about how to push code from the command line using a GitHub App. I ended adding some code to update the config with a user and email, so that user is the committer/author shown on GitHub. Here is the reference slack [convo](https://github.slack.com/archives/CD9PT6ZDY/p1543264767664500).

DEGoodmanWilson replied:
This seems a huge improvement 💃

rachmari commented:
@DEGoodmanWilson thanks for taking time to review this:

> There are some security issues that are pretty serious—it's easy to make this mistake! We should be leaders with security best practices, and the necessary changes are pretty easy to make (happy to help out if you need it!)

I absolutely agree. I definitely wanted to get some developers to take a  👀 at this for that very reason. If you'd like to contribute to this PR, you're more than welcome 😄. I'm not very well-versed with security best practices, so I'm learning as I go. 🙇‍♀️

runleonarun commented on advanced_server.rb:
```diff
@@ -0,0 +1,314 @@
+require 'logger'
+
+set :port, 3000
+
```
I had to add this code in line 9 to get smee to receive events:
`set :bind, '0.0.0.0'`

runleonarun commented on server.rb:
```diff
@@ -0,0 +1,160 @@
+require 'logger'
+
+set :port, 3000
+
```
I had to add this code in line 9 to get smee to receive events:
set :bind, '0.0.0.0'

swinton commented on README.md:
```diff
@@ -1 +1,13 @@
-# building-a-checks-api-ci-server
\ No newline at end of file
+This is the sample project he built by following the "[Building Your First GitHub App](https://developer.github.com/apps/building-your-first-github-app)" Quickstart guide on developer.github.com.
```
Seems there's a typo in the first sentence here. Should this be:

> This is the sample project built by following

Or even in the passive:

> Build this sample project by following

swinton commented on README.md:
```diff
@@ -1 +1,13 @@
\ No newline at end of file
+This is the sample project he built by following the "[Building Your First GitHub App](https://developer.github.com/apps/building-your-first-github-app)" Quickstart guide on developer.github.com.
+
+It consists of two different servers: `server.rb` (boilerplate) and `advanced_server.rb` (completed project).
```
What are your thoughts on progressing the evolution of `server.rb` through branches vs in different files on the same branch?

rachmari replied:
I'm still considering this because I love how elegant it would be to show the progress. But, we're talking about breaking out the template code (server.rb) into its own repository because it will serve as a template for future GitHub App guides, and we want to be as DRY as possible. I am keeping this in mind though until we settle on a solution.

swinton commented on README.md:
```diff
@@ -1 +1,13 @@
+
+* For the boilerplate project, enter `ruby server.rb` on the command line.
+
+- For the completed project, enter `ruby advanced_server.rb` on the command line.
```
Nit: avoid switching between `*` and `-` for lists

swinton commented on advanced_server.rb:
```diff
@@ -0,0 +1,314 @@
+
+class GHAapp < Sinatra::Application
+
+  # !!! DO NOT EVER USE HARD-CODED VALUES IN A REAL APP !!!
```
I think the example _is_ using an environment variable, so perhaps remove this line?

DEGoodmanWilson replied:
I had meant it as a warning: Look, do like we are doing, and please don't revert to the easy but insecure choice. But it does read like a disclaimer: Look, this is bad, please don't do this, I'm doing this only because it is convenient. Perhaps I could make that clearer.

rachmari replied:
I'll pull this out, because we do cover the use of environment variables for security reasons in the Quickstart guide.

swinton commented on advanced_server.rb:
```diff
@@ -0,0 +1,314 @@
+
+  # Your registered app must have a secret set. The secret is used to verify
+  # that webhooks are sent by GitHub.
+  WEBHOOK_SECRET = ENV['GITHUB_WEBHOOK_SECRET']
```
I think it would be helpful to collect all the environment variables needed to run the code in a `.env.example` file. Perhaps the `README` could include an instruction to create a `.env` from this `.env.example` template, and then the code updated to use [the `dotenv` pattern](https://github.com/bkeepers/dotenv) to load in the user's `.env` (which should be ignored in `.gitignore`.

rachmari replied:
I love this idea! I've updated the code to use dotenv, and I'll update the instructions in the guide to follow.

swinton commented on advanced_server.rb:
```diff
@@ -0,0 +1,314 @@
+      if @payload['check_run']['app']['id'].to_s === APP_IDENTIFIER
+        case @payload['action']
+        when 'created'
+          initiate_check_run
```
This could also happen _synchronously_ following `create_check_run`, rather then waiting for GitHub to send another webhook, the `created` webhook event handling could then be ignored.

DEGoodmanWilson replied:
I wanted to guarantee the timing of things on the server side, rather than assuming that GitHub will receive and act on the response—so the code would be more robust in the face of GH failures.

DEGoodmanWilson replied:
Also it demonstrates that in fact GitHub will essentially provide a receipt for the creation of a check_run, and that you can hook into that event.

swinton commented on advanced_server.rb:
```diff
@@ -0,0 +1,314 @@
+
+      clone_repository(full_repo_name, repository, head_sha)
+
+      @report = `rubocop '#{repository}/*' --format json` # --auto-correct`
```
Consider removing the comment at the end of this line if it's not needed:

```
# --auto-correct`
```

osowskit commented on advanced_server.rb:
```diff
@@ -0,0 +1,314 @@
+        end
+      end
+    end
+  end
```
I believe I needed to return some value to the POST request, `200` or `201`, to resolve an issue trying to return the check_run response to the a client.

rachmari replied:
I added a `status 200` to be returned from the POST request. 👍

osowskit commented on advanced_server.rb:
```diff
@@ -0,0 +1,314 @@
+          output: {
+            title: "Octoc Rubocop",
+            summary: summary,
+            details: details,
```
According to [the docs](https://developer.github.com/v3/checks/runs/#parameters-1), `details:` should probably be `text:`.

rachmari replied:
Oops! Yes, that is not the correct param name. Updated!

rachmari commented:
@swinton @DEGoodmanWilson @osowskit I haven't integrated all of your feedback yet, but I have address several items:

* We're now using dotenv 🎉 
* We're now using ruby-git to run git operations!! 
* We're now returning a `status 200` from the POST request. I wonder if I'm still missing return values in other methods.
* Fixed a bug in the parameter name for a checks API I was using and updated some parameter names for the changes that were made recently.

I tried adding the access token to the push command using the same format as the clone command, but got a failure. Looking into what went wrong there.

osowskit commented on advanced_server.rb:
```diff
@@ -0,0 +1,312 @@
+  before '/event_handler' do
+    get_payload_request(request)
+    verify_webhook_signature
+    authenticate_app
```
This will `404` for Check suite `created` webhooks for another app. The `if @payload['check_run']['app']['id'].to_s === APP_IDENTIFIER` check should get moved here.

rachmari replied:
@osowskit ah I see the problem there. Should we move the `authenticate_app` and `authenticate_installation` methods into each `when` statement below? We'll have to duplicate the code for each `when` statement, but new tokens would only be created when they had passed all of the checks.

rachmari replied:
Circling back to this comment, I realized that this update I added before the `authenticate_app` method wouldn't work:

```
halt 400 unless @payload[request.env['HTTP_X_GITHUB_EVENT']]['app']['id'].to_s === APP_IDENTIFIER
```

The APP_ID is inside of the event object, so it didn't seem very clean to check for more than one expected event. I think the app will `401` when the payload is for another app in the `verify_webhook_signature` before the `authenticate_app` method is called. I'm going to test how this works in the AM when I have two apps installed on the same repository creating check runs. If you see a good workaround, I'm all ears.



```
{
  "action": "completed",
  "check_suite": {
        "app": {
                    "id": 17104
         }
    }
}
```

DEGoodmanWilson commented:
I have added two small changes—one just a matter of convention (adding a `!` to method names that can abort early), but the other to demonstrate the kind of input validation I mentioned earlier. If you don't like, or just don't want, please feel free to back either of these out! But I think it is a good idea to demonstrate validation like this.

DEGoodmanWilson commented on advanced_server.rb:
```diff
@@ -0,0 +1,319 @@
+      if @output["summary"]["offense_count"] == 0
+        conclusion = 'success'
+      else
+        conclusion = 'neutral'
```
I'm curious why "neutral" and not "failure"?

rachmari replied:
@DEGoodmanWilson that's a good question, and maybe we should change it to "failure." I did this for two reasons:
1. To demonstrate that a "neutral" status could be used (which is not a good demonstration unless we explain this in the quickstart - noted 😄)
2. I chose "neutral" because linters can have very specific rules that might act as more of guidelines rather than rigid rules. However, because the fix button only allows fixing all of the errors, it might be better to use "failure" here.

DEGoodmanWilson commented on advanced_server.rb:
```diff
@@ -217,15 +217,20 @@ def take_requested_action
     end
 
     def clone_repository(full_repo_name, repository, head_sha, head_branch=nil)
+      print  "https://x-access-token:#{@installation_token.to_s}@github.com/#{full_repo_name}.git\n"
```
Oh crap, this line shouldn't have made it into the commit!

DEGoodmanWilson commented:
Howdy! Where does this stand at the moment? I've rather lost the thread of things here.

rachmari commented:
@DEGoodmanWilson @swinton @osowskit 👋I think I've made all the updates I needed to make based on your feedback. There are still 3 unresolved comments/conversations, but I'm not sure that any action is needed:

* Do we need a token or some kind of authorization to push the code: [comment link](https://github.com/github-developer-docs-samples/building-a-checks-api-ci-server/pull/1#discussion_r210752742)
* Do we need to update the `verify_webhook_signature` method so it doesn't `401` when other app's checks are created: [comment link](https://github.com/github-developer-docs-samples/building-a-checks-api-ci-server/pull/1#discussion_r214545094)
* Should we change `neutral` to `failure` for the check run conclusion - Note I added some descripton around why I used `neutral` in the Quickstart and describe that you can use `failure` instead. Here's the [PR with that change](https://github.com/github/internal-developer.github.com/pull/3851/files#diff-bbdb48f646921127899f499ecd07357fR443): [comment link](https://github.com/github-developer-docs-samples/building-a-checks-api-ci-server/pull/1#discussion_r216702062). 

Any thoughts on ☝️topics from anyone would be great. 

There was one change that might throw you for a loop if you're coming back to this review. The names of these files changed:
`advanced_server.rb` -> `server.rb`
`server.rb` -> `template_server.rb`.

We'd like to keep the `template_server.rb` file the same across all Quickstarts as the base template, so this naming convention seemed to make more sense. Open to suggestions there.

Finally, I'd love to get another review of `server.rb` and any other files you have time for. 🙇‍♀️ 💚

DEGoodmanWilson commented on server.rb:
```diff
@@ -0,0 +1,330 @@
+    def take_requested_action
+      full_repo_name = @payload['repository']['full_name']
+      repository     = @payload['repository']['name']
+      head_branch    = @payload['check_run']['check_suite']['head_branch']
```
This turns out to be a problem—When I open a PR, requesting to merge a branch `test` into `master`, like this one:
https://github.com/DEGoodmanWilson/internal-sandbox/pull/1/checks
The value of `head_branch` is `master`. So when the "fix changes" button gets clicked, the CI server is pushing the changes into the _wrong branch_ :sob: It's also checking out the wrong branch for running Rubocop on.

We want to check out, and push our changes onto the `test` branch, but I can't see where that information is available to us. All we have is the SHA1 for the commit that we are supposed to be checking, which is in `@payload['check_run']['check_suite']['head_sha']`. _this_ value is correct: On my PR, it resolves to `5530c5a92d3d4649329aa11cccff61aa26128569`, which is indeed the SHA for `$HEAD` on the `test` branch.

DEGoodmanWilson replied:
Sorry about that! It might also be that the commit history is _seriously_ messed up. If it is working for you, then I am inclined to just say my situation is a fluke, and just ignore it.

DEGoodmanWilson replied:
Oh, but I did add you as a collaborator :D

rachmari commented:
@DEGoodmanWilson [Rubcop has a few different severities](https://rubocop.readthedocs.io/en/latest/configuration/#severity): refactor, convention, warning, error and fatal.

We could set a status of `neutral` for refactor, convention, and warning and a status of `failure` for error and fatal.

💭 :octocat:

rachmari commented:
@swinton @DEGoodmanWilson @osowskit I think we can 🚢 this! Here is the latest summary:

* I don't think any updates are required to the `verify_webhook_signature` method related to this [comment](https://github.com/github-developer-docs-samples/building-a-checks-api-ci-server/pull/1#discussion_r214545094). I've tested this code with two applications running on two different ports on the same repository both creating and receiving check runs. There were no 401 errors and the check runs were created successfully. Here is the flow that's happening:
   - A `completed` check suite and `created` check run event is sent to each app on a repository. The app verifies the webhook, which succeeds because the webhook is valid. 
    - The app then authenticates for the installation. This is successful too because the installation ID used is still correct.
    -  The check run event handler will check the check run app id to see if it matches the app id. If it doesn't, the program execution ends without an error.
* We chatted in [this comment](https://github.com/github-developer-docs-samples/building-a-checks-api-ci-server/pull/1#issuecomment-442671346) about using two different statuses for the different kinds of Rubocop errors. I definitely want to add this, but my suggestion is that we ship this first iteration and implement that update in the next revision. I can create an issue to track that update. 

If there are no issues with the suggestion ☝️, I'll go ahead and merge this!

DEGoodmanWilson commented:
https://media.giphy.com/media/wi8Ez1mwRcKGI/giphy.gif

swinton commented on .env-example:
```diff
@@ -0,0 +1,3 @@
+GITHUB_PRIVATE_KEY=""
```
Are the quotes necessary here? 🤔

rachmari replied:
@swinton the quotes are necessary. The quickstart guide explains why. Quotes are needed for multi-line variables, for the PEM key.

DEGoodmanWilson commented on server.rb:
```diff
@@ -0,0 +1,328 @@
+      )
+
+      # You requested the creation of a check run from GitHub. Now, you'll wait
+      # to get confirmation from GitHub that it was created before starting CI.
```
Makes testing _so_ much easier this way, by just re-sending the created event (from Smee, _e.g._).

rachmari commented on server.rb:
```diff
@@ -0,0 +1,328 @@
+
+        # Sets your commit username and email address
+        @git.config('user.name', ENV['GITHUB_APP_USER_NAME'])
+        @git.config('user.email', ENV['GITHUB_APP_USER_EMAIL'])
```
@swinton yes, this is the GitHub App email and user. I couldn't find a better way to ensure the commit was made by the app versus another user who has credentials set on the server.

rachmari commented on server.rb:
```diff
@@ -0,0 +1,328 @@
+            title: 'Octo RuboCop',
+            summary: summary,
+            text: text,
+            annotations: annotations
```
@swinton good catch! I'll make that update.

rachmari replied:
I think for this version, we'll just add a comment that it will be truncated to the first 50. I'll create an issue to capture adding batching.

swinton commented on README.md:
```diff
@@ -1 +1,21 @@
+* `template_server.rb` (GitHub App template code)
+* `server.rb` (completed project)
+
+To learn how to set up a template GitHub App, follow the "[Configuring a GitHub App](https://developer.github.com/apps/quickstart-guide/configuring-ast-github-app)" Quickstart guide on developer.github.com.
```
Does this guide include [a manifest](https://developer.github.com/apps/building-github-apps/creating-github-apps-from-a-manifest/) to help folks get started with the GitHub App creation part? What are your thoughts on including it here?

rachmari replied:
@swinton ah yes!! I had planned to add GitHub Manifests to quickstarts, but it's going to be in a subsequent iteration. I'll create an issue to track that!

swinton commented on server.rb:
```diff
@@ -153,6 +159,10 @@ def initiate_check_run
 
           # Parse each offense to get details and location
           file['offenses'].each do |offense|
+            # Limit the number of annotations to 50
```
To avoid having to update `max_annotations` in each iteration, what do you think about just taking a slice of the first 50 `offenses` outside of the loop? The following will work even if there are less than 50 offenses.

```ruby
# process first 50 offenses
file['offenses'][0, 50].each do |offense|
  # process offense ...
done
```

rachmari replied:
I like this much better! 👍

rachmari replied:
@swinton thinking about this more, will this work if there is more than one file. The loop iterates first through the files then through the offenses in each file. Will this allow 50 offenses per file?

swinton replied:
> this work if there is more than one file. The loop iterates first through the files then through the offenses in each file. Will this allow 50 offenses per file?

Ah, nope this won't work, I hadn't appreciated that there are many files with potentially many offenses. Good catch! ⚾️

rachmari replied:
@swinton thanks for confirming. I think all the changes are in. ⚡️ I'm going to merge this. 🎉 ✨ 

I'll merge the quickstart guides on Monday after I've moved this repo to github-developer. I'm hoping for a pretty short turnaround to supporting more than 50 annotations.
